File: src/orchestrator.py
"""Session orchestrator — coordinates RTMS → face detection → Moveris pipeline.

One ``SessionOrchestrator`` manages all active meeting sessions.  For each
meeting a ``_Session`` is created that:

1. Connects an ``RTMSClient`` to the Zoom RTMS stream.
2. Routes decoded video frames to per-participant asyncio queues.
3. For each participant, runs a background task that filters frames,
   detects faces with MediaPipe, accumulates 10 quality crops, then
   submits them to the Moveris ``fast-check-crops`` API.
4. Stores ``LivenessResult`` objects in the injected ``ResultStore``.
"""

import asyncio
import logging
import time
from datetime import UTC, datetime

import numpy as np

from src.config import Settings
from src.moveris.client import MoverisClient, MoverisError, MoverisResponse
from src.results import InMemoryResultStore, LivenessResult, ResultStore
from src.rtms.client import RTMSClient
from src.video.face_detector import FaceDetector
from src.video.frame_selector import is_quality_frame

logger = logging.getLogger(__name__)

_MIN_CROPS = 10  # frames required by fast-check-crops
_FRAME_TIMEOUT_S = 30.0  # max seconds to wait for the next frame
_QUEUE_MAXSIZE = 100  # drop frames if a participant queue fills up


class TooManySessions(Exception):
    """Raised when ``MAX_CONCURRENT_SESSIONS`` would be exceeded."""


class SessionOrchestrator:
    """Top-level coordinator for all active RTMS sessions.

    Usage::

        orchestrator = SessionOrchestrator(settings, result_store)
        await orchestrator.start_session(meeting_uuid, rtms_stream_id, server_urls)
        # … frames processed automatically …
        await orchestrator.stop_session(meeting_uuid)
        await orchestrator.close()  # shut down all remaining sessions
    """

    def __init__(
        self,
        settings: Settings,
        result_store: ResultStore | None = None,
    ) -> None:
        self._settings = settings
        self._result_store: ResultStore = result_store or InMemoryResultStore()
        self._sessions: dict[str, _Session] = {}
        self._lock = asyncio.Lock()

    @property
    def active_session_count(self) -> int:
        return len(self._sessions)

    @property
    def result_store(self) -> ResultStore:
        return self._result_store

    async def start_session(
        self,
        meeting_uuid: str,
        rtms_stream_id: str,
        server_urls: str,
    ) -> None:
        """Begin streaming and liveness analysis for a meeting.

        Raises:
            TooManySessions: If ``MAX_CONCURRENT_SESSIONS`` would be exceeded.
        """
        async with self._lock:
            if len(self._sessions) >= self._settings.max_concurrent_sessions:
                raise TooManySessions(
                    f"Cannot start session {meeting_uuid}: "
                    f"max {self._settings.max_concurrent_sessions} concurrent sessions"
                )
            if meeting_uuid in self._sessions:
                logger.warning(
                    "Session already active — ignoring duplicate: %s", meeting_uuid
                )
                return

            await self._result_store.create_session(meeting_uuid)
            await self._result_store.set_session_state(meeting_uuid, "processing")

            session = _Session(
                meeting_uuid=meeting_uuid,
                rtms_stream_id=rtms_stream_id,
                server_urls=server_urls,
                settings=self._settings,
                result_store=self._result_store,
            )
            self._sessions[meeting_uuid] = session

        await session.start()
        logger.info("Session started: meeting=%s", meeting_uuid)

    async def stop_session(self, meeting_uuid: str) -> None:
        """Gracefully stop an active session and mark it complete."""
        async with self._lock:
            session = self._sessions.pop(meeting_uuid, None)

        if session is None:
            logger.debug("stop_session called for unknown meeting: %s", meeting_uuid)
            return

        await session.close()
        await self._result_store.set_session_state(meeting_uuid, "complete")
        logger.info("Session stopped: meeting=%s", meeting_uuid)

    async def close(self) -> None:
        """Stop all active sessions."""
        async with self._lock:
            sessions = list(self._sessions.items())
            self._sessions.clear()

        for meeting_uuid, session in sessions:
            await session.close()
            await self._result_store.set_session_state(meeting_uuid, "complete")

        if sessions:
            logger.info(
                "SessionOrchestrator shut down (%d sessions closed)", len(sessions)
            )


# ---------------------------------------------------------------------------
# Internal: per-meeting session
# ---------------------------------------------------------------------------


class _Session:
    def __init__(
        self,
        meeting_uuid: str,
        rtms_stream_id: str,
        server_urls: str,
        settings: Settings,
        result_store: ResultStore,
    ) -> None:
        self._meeting_uuid = meeting_uuid
        self._settings = settings
        self._result_store = result_store
        self._participant_queues: dict[str, asyncio.Queue] = {}
        self._participant_tasks: dict[str, asyncio.Task] = {}
        self._rtms = RTMSClient(
            settings=settings,
            meeting_uuid=meeting_uuid,
            rtms_stream_id=rtms_stream_id,
            server_urls=server_urls,
            on_frame=self._on_frame,
        )

    async def start(self) -> None:
        await self._rtms.start()

    async def close(self) -> None:
        await self._rtms.close()

        for task in self._participant_tasks.values():
            task.cancel()

        if self._participant_tasks:
            await asyncio.gather(
                *self._participant_tasks.values(), return_exceptions=True
            )

        self._participant_tasks.clear()
        self._participant_queues.clear()

    async def _on_frame(
        self,
        frame: np.ndarray,
        user_id: int,
        user_name: str,
        timestamp_ms: int,
    ) -> None:
        """Route a decoded video frame to the correct participant task."""
        if user_id == 0:
            return  # SDK metadata unavailable for this frame

        participant_id = str(user_id)

        if participant_id not in self._participant_queues:
            queue: asyncio.Queue[np.ndarray | None] = asyncio.Queue(
                maxsize=_QUEUE_MAXSIZE
            )
            self._participant_queues[participant_id] = queue
            task = asyncio.create_task(
                self._process_participant(participant_id, queue),
                name=f"participant-{self._meeting_uuid[:8]}-{participant_id}",
            )
            self._participant_tasks[participant_id] = task
            logger.info(
                "Spawned participant task — meeting=%s participant=%s",
                self._meeting_uuid,
                participant_id,
            )

        queue = self._participant_queues[participant_id]
        try:
            queue.put_nowait(frame)
        except asyncio.QueueFull:
            logger.debug(
                "Frame queue full — dropping frame for participant %s", participant_id
            )

    async def _process_participant(
        self,
        participant_id: str,
        queue: asyncio.Queue,
    ) -> None:
        """Collect quality face crops and submit to Moveris."""
        detector = FaceDetector()
        crops: list[str] = []
        frames_seen = 0
        start = time.monotonic()

        try:
            async with MoverisClient(api_key=self._settings.moveris_api_key) as moveris:
                while len(crops) < _MIN_CROPS:
                    try:
                        frame = await asyncio.wait_for(
                            queue.get(), timeout=_FRAME_TIMEOUT_S
                        )
                    except asyncio.TimeoutError:
                        logger.warning(
                            "Frame timeout for participant %s in meeting %s "
                            "after %d frames seen (%d/%d crops collected)",
                            participant_id,
                            self._meeting_uuid,
                            frames_seen,
                            len(crops),
                            _MIN_CROPS,
                        )
                        break

                    if frame is None:  # sentinel from close()
                        break

                    frames_seen += 1

                    if not is_quality_frame(frame):
                        continue

                    image_b64 = detector.detect(frame)
                    if image_b64 is None:
                        continue

                    crops.append(image_b64)

                if len(crops) < _MIN_CROPS:
                    await self._store_error(
                        participant_id,
                        "insufficient_frames",
                        frames_seen,
                        start,
                    )
                    return

                await self._call_moveris(
                    moveris, participant_id, crops[:_MIN_CROPS], frames_seen, start
                )

        except asyncio.CancelledError:
            logger.debug(
                "Participant task cancelled — meeting=%s participant=%s",
                self._meeting_uuid,
                participant_id,
            )
            raise
        except Exception as exc:
            logger.exception(
                "Unexpected error processing participant %s in meeting %s: %s",
                participant_id,
                self._meeting_uuid,
                exc,
            )
            await self._store_error(participant_id, str(exc), frames_seen, start)
        finally:
            detector.close()

    async def _call_moveris(
        self,
        client: MoverisClient,
        participant_id: str,
        crops: list[str],
        frames_seen: int,
        start: float,
    ) -> None:
        try:
            response: MoverisResponse = await client.check_crops(crops)
            result = LivenessResult(
                meeting_uuid=self._meeting_uuid,
                participant_id=participant_id,
                verdict=response.verdict,
                score=response.score,
                real_score=response.real_score,
                fake_score=response.fake_score,
                confidence=response.confidence,
                processing_ms=int((time.monotonic() - start) * 1000),
                frames_processed=frames_seen,
                session_id="",
                completed_at=datetime.now(UTC),
            )
            await self._result_store.set_result(
                self._meeting_uuid, participant_id, result
            )
            logger.info(
                "Liveness result — meeting=%s participant=%s verdict=%s score=%.1f",
                self._meeting_uuid,
                participant_id,
                response.verdict,
                response.score,
            )
        except MoverisError as exc:
            logger.error(
                "Moveris API error for participant %s in meeting %s: %s",
                participant_id,
                self._meeting_uuid,
                exc,
            )
            await self._store_error(participant_id, str(exc), frames_seen, start)

    async def _store_error(
        self,
        participant_id: str,
        error: str,
        frames_seen: int,
        start: float,
    ) -> None:
        await self._result_store.set_result(
            self._meeting_uuid,
            participant_id,
            LivenessResult(
                meeting_uuid=self._meeting_uuid,
                participant_id=participant_id,
                verdict="error",
                score=0.0,
                real_score=0.0,
                fake_score=0.0,
                confidence=0.0,
                processing_ms=int((time.monotonic() - start) * 1000),
                frames_processed=frames_seen,
                session_id="",
                completed_at=datetime.now(UTC),
                error=error,
            ),
        )

File: src/orchestrator.ts
import {
  LivenessApiError,
  LivenessClient,
  generateSessionId,
  type CapturedFrame,
} from "@moveris/shared";
import type { Config } from "./config.js";
import type { ResultStore } from "./results.js";
import { RTMSClient } from "./rtms-client.js";
import { H264BatchDecoder } from "./h264-batch-decoder.js";
import type { ParticipantResult } from "./types.js";

/** How often to check if the batch decoder is ready (ms). */
const ACCUMULATION_CHECK_INTERVAL_MS = 250;
/** Max time to wait for H264 data before giving up (ms). */
const ACCUMULATION_TIMEOUT_MS = 30_000;

export class TooManySessions extends Error {}

interface ParticipantState {
  decoder: H264BatchDecoder;
  done: boolean;
  userName: string;
  checkInterval: ReturnType<typeof setInterval>;
  timeout: ReturnType<typeof setTimeout>;
}

/**
 * Progress callback — reports accumulation time instead of frame count.
 * framesCollected = elapsed seconds (e.g., 2.1)
 * framesNeeded = target seconds (e.g., 4.0)
 */
export type ProgressCallback = (
  meetingUuid: string,
  participantId: string,
  userName: string,
  framesCollected: number,
  framesNeeded: number,
) => void;

/** Lifecycle stage updates pushed to the sidebar. */
export type StageCallback = (
  meetingUuid: string,
  participantId: string | null,
  userName: string | null,
  stage: "connected" | "recording" | "decoding" | "analyzing",
) => void;

export type ResultCallback = (
  meetingUuid: string,
  participantId: string,
  result: ParticipantResult,
) => void;

interface PendingSession {
  meetingUuid: string;
  apiKey?: string;
  createdAt: Date;
}

export class SessionOrchestrator {
  private sessions = new Map<string, Session>();
  private pendingSessions = new Map<string, PendingSession>();
  private config: Config;
  private resultStore: ResultStore;
  private defaultLivenessClient: LivenessClient | null = null;
  onProgress: ProgressCallback | null = null;
  onStage: StageCallback | null = null;
  onResult: ResultCallback | null = null;

  constructor(config: Config, resultStore: ResultStore) {
    this.config = config;
    this.resultStore = resultStore;
    if (config.MOVERIS_API_KEY) {
      this.defaultLivenessClient = new LivenessClient({
        apiKey: config.MOVERIS_API_KEY,
        enableRetry: true,
      });
    }
  }

  get activeSessionCount(): number {
    return this.sessions.size;
  }

  registerPendingSession(meetingUuid: string, apiKey?: string): void {
    this.pendingSessions.set(meetingUuid, {
      meetingUuid,
      apiKey,
      createdAt: new Date(),
    });
    console.log(`Pending session registered: meeting=${meetingUuid}`);
  }

  hasPendingSession(meetingUuid: string): boolean {
    return this.pendingSessions.has(meetingUuid);
  }

  consumePendingSession(meetingUuid: string): string | undefined {
    const pending = this.pendingSessions.get(meetingUuid);
    if (!pending) return undefined;
    this.pendingSessions.delete(meetingUuid);
    return pending.apiKey;
  }

  startSession(meetingUuid: string, rtmsPayload: Record<string, any>, apiKey?: string): void {
    if (this.sessions.size >= this.config.MAX_CONCURRENT_SESSIONS) {
      throw new TooManySessions(
        `Cannot start session ${meetingUuid}: max ${this.config.MAX_CONCURRENT_SESSIONS} concurrent sessions`,
      );
    }
    if (this.sessions.has(meetingUuid)) {
      console.log(`Session already active — ignoring duplicate: ${meetingUuid}`);
      return;
    }

    let livenessClient: LivenessClient;
    if (apiKey) {
      livenessClient = new LivenessClient({ apiKey, enableRetry: true });
    } else if (this.defaultLivenessClient) {
      livenessClient = this.defaultLivenessClient;
    } else {
      console.error(`No API key for session ${meetingUuid} and no default key configured`);
      return;
    }

    this.resultStore.createSession(meetingUuid);
    this.resultStore.setSessionState(meetingUuid, "processing");

    const session = new Session(
      meetingUuid,
      rtmsPayload,
      this.resultStore,
      livenessClient,
      (reason: string) => this.onSessionError(meetingUuid, reason),
      this.onProgress,
      this.onStage,
      this.onResult,
    );
    this.sessions.set(meetingUuid, session);
    session.start();
    console.log(`Session started: meeting=${meetingUuid}`);
  }

  stopSession(meetingUuid: string): void {
    const session = this.sessions.get(meetingUuid);
    if (!session) return;
    this.sessions.delete(meetingUuid);
    session.close();
    this.resultStore.setSessionState(meetingUuid, "complete");
    console.log(`Session stopped: meeting=${meetingUuid}`);
  }

  close(): void {
    for (const [uuid, session] of this.sessions) {
      session.close();
      this.resultStore.setSessionState(uuid, "complete");
    }
    const count = this.sessions.size;
    this.sessions.clear();
    if (count > 0) {
      console.log(`SessionOrchestrator shut down (${count} sessions closed)`);
    }
  }

  private onSessionError(meetingUuid: string, reason: string): void {
    const session = this.sessions.get(meetingUuid);
    if (!session) return;
    this.sessions.delete(meetingUuid);
    session.close();
    this.resultStore.setSessionState(meetingUuid, "error");
    console.error(`Session error: meeting=${meetingUuid} reason=${reason}`);
  }
}

class Session {
  private rtms: RTMSClient;
  private participants = new Map<string, ParticipantState>();
  private meetingUuid: string;
  private resultStore: ResultStore;
  private livenessClient: LivenessClient;
  private onSessionError: (reason: string) => void;
  private onProgress: ProgressCallback | null;
  private onStage: StageCallback | null;
  private onResult: ResultCallback | null;

  constructor(
    meetingUuid: string,
    rtmsPayload: Record<string, any>,
    resultStore: ResultStore,
    livenessClient: LivenessClient,
    onSessionError: (reason: string) => void,
    onProgress: ProgressCallback | null,
    onStage: StageCallback | null,
    onResult: ResultCallback | null,
  ) {
    this.meetingUuid = meetingUuid;
    this.resultStore = resultStore;
    this.livenessClient = livenessClient;
    this.onSessionError = onSessionError;
    this.onProgress = onProgress;
    this.onStage = onStage;
    this.onResult = onResult;
    this.rtms = new RTMSClient(
      meetingUuid,
      rtmsPayload,
      // Raw H264 chunks — no decoding at the RTMS layer
      (h264Data, userId, userName, _timestampMs) => {
        this.onH264Chunk(h264Data, userId, userName);
      },
      () => {
        // RTMS joined — notify sidebar immediately
        this.onStage?.(this.meetingUuid, null, null, "connected");
      },
      (userId, userName) => {
        // Video data first seen for this user — show them instantly
        const participantId = String(userId);
        const displayName = userName || participantId;
        console.log(`Participant detected — meeting=${this.meetingUuid} participant=${participantId} name=${displayName}`);
        this.onStage?.(this.meetingUuid, participantId, displayName, "recording");
        // Report 0 / 4 seconds initially
        this.onProgress?.(this.meetingUuid, participantId, displayName, 0, 4);
      },
      (reason) => {
        this.onSessionError(`rtms_join_failed (reason=${reason})`);
      },
      (reason) => {
        this.onSessionError(`rtms_disconnected (reason=${reason})`);
      },
    );
  }

  start(): void {
    this.rtms.start();
  }

  close(): void {
    this.rtms.close();
    for (const [, state] of this.participants) {
      clearInterval(state.checkInterval);
      clearTimeout(state.timeout);
      if (!state.done) {
        state.decoder.cancel();
      }
    }
    this.participants.clear();
  }

  /**
   * Receives raw H264 chunks from RTMSClient.
   * Feeds them into the per-participant batch decoder.
   */
  private onH264Chunk(h264Data: Buffer, userId: number, userName: string): void {
    if (userId === 0) return;

    const participantId = String(userId);
    let state = this.participants.get(participantId);

    if (!state) {
      const displayName = userName || participantId;
      const decoder = new H264BatchDecoder();

      // Periodically check if the batch decoder has accumulated enough data
      const checkInterval = setInterval(() => {
        this.checkAccumulation(participantId);
      }, ACCUMULATION_CHECK_INTERVAL_MS);

      // Overall timeout: if we never get enough data
      const timeout = setTimeout(() => {
        this.onParticipantTimeout(participantId);
      }, ACCUMULATION_TIMEOUT_MS);

      state = {
        decoder,
        done: false,
        userName: displayName,
        checkInterval,
        timeout,
      };
      this.participants.set(participantId, state);
      console.log(`Spawned batch collector — meeting=${this.meetingUuid} participant=${participantId} name=${displayName}`);
    }

    if (state.done) return;

    // Feed raw H264 data into the batch decoder
    state.decoder.feed(h264Data);
  }

  /**
   * Called periodically to check if we've accumulated enough H264 data.
   * Pushes time-based progress updates to the sidebar.
   */
  private checkAccumulation(participantId: string): void {
    const state = this.participants.get(participantId);
    if (!state || state.done) return;

    const elapsedMs = state.decoder.getElapsedMs();
    const elapsedSec = elapsedMs / 1000;

    // Push time-based progress (seconds accumulated / 4 seconds target)
    this.onProgress?.(this.meetingUuid, participantId, state.userName, elapsedSec, 4);

    // Check if ready to decode
    if (state.decoder.isReady()) {
      state.done = true;
      clearInterval(state.checkInterval);
      clearTimeout(state.timeout);

      console.log(`Batch ready — meeting=${this.meetingUuid} participant=${participantId} (${state.decoder.getTotalBytes()} bytes accumulated)`);

      this.decodeBatch(participantId, state).catch((err) => {
        console.error(`Batch decode error for participant ${participantId}: ${err}`);
      });
      return;
    }

    // Check for inactivity timeout
    if (state.decoder.isTimedOut()) {
      state.done = true;
      clearInterval(state.checkInterval);
      clearTimeout(state.timeout);
      state.decoder.cancel();

      console.warn(`Inactivity timeout for participant ${participantId} in meeting ${this.meetingUuid}`);

      const errorResult: ParticipantResult = {
        meetingUuid: this.meetingUuid,
        participantId,
        result: null,
        completedAt: new Date(),
        error: "inactivity_timeout",
      };
      this.resultStore.setResult(this.meetingUuid, participantId, errorResult);
      this.onResult?.(this.meetingUuid, participantId, errorResult);
    }
  }

  /**
   * Decode the accumulated H264 batch for a participant,
   * then submit 10 consecutive frames to Moveris.
   */
  private async decodeBatch(participantId: string, state: ParticipantState): Promise<void> {
    // Notify sidebar: decoding
    this.onStage?.(this.meetingUuid, participantId, state.userName, "decoding");

    try {
      const { frames: pngFrames } = await state.decoder.decode();

      console.log(`Batch decoded — meeting=${this.meetingUuid} participant=${participantId} frames=${pngFrames.length}`);

      // Build CapturedFrame objects from the 10 consecutive PNG frames
      const capturedFrames: CapturedFrame[] = pngFrames.map((png, i) => ({
        index: i,
        timestampMs: Date.now(),
        pixels: png.toString("base64"),
      }));

      // Notify sidebar: analyzing (submitting to Moveris)
      this.onStage?.(this.meetingUuid, participantId, state.userName, "analyzing");

      await this.submitToMoveris(participantId, capturedFrames);
    } catch (err) {
      console.error(`Batch decode failed for participant ${participantId}: ${err}`);

      const errorResult: ParticipantResult = {
        meetingUuid: this.meetingUuid,
        participantId,
        result: null,
        completedAt: new Date(),
        error: `decode_error: ${err}`,
      };
      this.resultStore.setResult(this.meetingUuid, participantId, errorResult);
      this.onResult?.(this.meetingUuid, participantId, errorResult);
    }
  }

  private async submitToMoveris(participantId: string, frames: CapturedFrame[]): Promise<void> {
    try {
      const result = await this.livenessClient.fastCheck(frames, {
        sessionId: generateSessionId(),
        source: "live",
      });

      const participantResult: ParticipantResult = {
        meetingUuid: this.meetingUuid,
        participantId,
        result,
        completedAt: new Date(),
      };

      this.resultStore.setResult(this.meetingUuid, participantId, participantResult);
      this.onResult?.(this.meetingUuid, participantId, participantResult);
      console.log(
        `Liveness result — meeting=${this.meetingUuid} participant=${participantId} verdict=${result.verdict} score=${result.score}`,
      );
    } catch (err) {
      let errorCode = String(err);

      if (err instanceof LivenessApiError) {
        errorCode = err.code;
        if (err.code === "invalid_key") {
          console.error(`Moveris API key is invalid — check MOVERIS_API_KEY`);
        } else if (err.code === "insufficient_credits") {
          console.error(`Moveris account has insufficient credits`);
        } else if (err.code === "rate_limit_exceeded") {
          console.error(`Moveris rate limit exceeded for participant ${participantId}`);
        } else {
          console.error(`Moveris API error for participant ${participantId}: [${err.code}] ${err.message}`);
        }
      } else {
        console.error(`Moveris API error for participant ${participantId}: ${err}`);
      }

      const errorResult: ParticipantResult = {
        meetingUuid: this.meetingUuid,
        participantId,
        result: null,
        completedAt: new Date(),
        error: errorCode,
      };

      this.resultStore.setResult(this.meetingUuid, participantId, errorResult);
      this.onResult?.(this.meetingUuid, participantId, errorResult);
    }
  }

  private onParticipantTimeout(participantId: string): void {
    const state = this.participants.get(participantId);
    if (!state || state.done) return;
    state.done = true;
    clearInterval(state.checkInterval);
    state.decoder.cancel();

    console.log(
      `Accumulation timeout for participant ${participantId} in meeting ${this.meetingUuid} (${state.decoder.getTotalBytes()} bytes accumulated)`,
    );

    const errorResult: ParticipantResult = {
      meetingUuid: this.meetingUuid,
      participantId,
      result: null,
      completedAt: new Date(),
      error: "accumulation_timeout",
    };

    this.resultStore.setResult(this.meetingUuid, participantId, errorResult);
    this.onResult?.(this.meetingUuid, participantId, errorResult);
  }
}
